Charts: Add bar list chart

chihsuan opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes CHARTS-7


## Proposed changes:
<!--- Explain what functional changes your PR includes -->

This PR introduces a new `BarListChart` component (Open to suggestions for better naming.) along with comprehensive Storybook stories and automated tests. 

The component is designed to support the following use cases in WooCommerce Analytics:

<img width="310" alt="Screenshot 2025-06-04 at 17 56 57" src="https://github.com/user-attachments/assets/49586c25-12d0-40cc-ba27-da9973b2fd28" />
<img width="340" alt="Screenshot 2025-06-04 at 17 56 49" src="https://github.com/user-attachments/assets/7ee38636-75a5-4281-8cc4-2358368c6258" />

This implementation extends the existing `BarChart` component and leverages the `axis` children prop to render labels and values in the same row with custom components.

**BarListChart Component****:
- Renders a horizontal bar list chart using the existing `BarChart` as a base.
- Supports both single and multi-series data.
- Allows custom label and value renderers via `options.labelComponent` and `options.valueComponent`.
- Accepts custom formatters for labels and values.

🛠️ Refactors & Improvements
- Updated `BarChart` and related types to improve scale option flexibility and axis rendering.
- Fix group padding issue in BarChart and BarListChart
- Improve `withResponsive` type definition 


### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->


1 Go to charts folder
```sh
cd projects/js-packages/charts/
```

2. Run Storybook app instance
```sh
npm run storybook
```

3. http://localhost:50240/?path=/story/js-packages-charts-types-bar-list-chart--default

<img width="450" alt="Screenshot 2025-06-04 at 18 02 39" src="https://github.com/user-attachments/assets/6edbad20-3b41-426a-9df1-3005fa24bb6d" />
<img width="450" alt="Screenshot 2025-06-04 at 18 02 42" src="https://github.com/user-attachments/assets/b9710b13-aa50-46d9-a73c-bbebe6a611c6" />
<img width="450" alt="Screenshot 2025-06-04 at 18 02 46" src="https://github.com/user-attachments/assets/9d8c79b6-e042-4426-8a09-40fbefc2fe40" />
<img width="450" alt="Screenshot 2025-06-04 at 18 02 54" src="https://github.com/user-attachments/assets/94f54ac6-50fd-4f52-84f0-259937447b3a" />

chihsuan committed:
Enhance chart types and BarChart component

- Updated `types.ts` to include new properties for `AxisOptions` and `ScaleOptions`, allowing for more flexible axis rendering and scale configuration.
- Refactored `BarChart` component to ensure axis rendering is consistent and correctly utilizes the updated options.
- Renamed `getInnerPadding` to `getGroupPadding` for clarity, aligning with the new scale properties.
- Improved documentation for scale padding properties to enhance developer understanding.

chihsuan committed:
Refactor withResponsive to exclude 'options' from BaseChartProps

- Updated the type definition of the `withResponsive` function to exclude the 'options' property from `BaseChartProps`, allowing each chart to define its own options type. This change enhances type safety and flexibility in chart component configurations.

chihsuan committed:
Add BarListChart component and related stories, types, and tests

- Introduced the BarListChart component for rendering horizontal bar charts with customizable labels and values.
- Created index and story files for the BarListChart, including examples for default, multi-series, and custom label/value components.
- Added sample data for testing and demonstration purposes.
- Implemented unit tests to ensure proper rendering and functionality of the BarListChart with various data scenarios.

chihsuan committed:
changelog

chihsuan committed:
Export bar list chart

chihsuan commented on projects/js-packages/charts/src/components/bar-chart/bar-chart.tsx:
```diff
@@ -142,6 +140,9 @@ const BarChart: FC< BarChartProps > = ( {
 				</AnimatedBarGroup>
 
+				<AnimatedAxis { ...chartOptions.axis.x } />
+				<AnimatedAxis { ...chartOptions.axis.y } />
```
Placed the `AnimatedAxis` components after the Bar component to ensure the axes are always visible.

chihsuan commented on projects/js-packages/charts/src/components/bar-chart/use-bar-chart-options.ts:
```diff
@@ -38,7 +38,7 @@ export function useBarChartOptions(
 			type: 'band' as const,
 			padding: 0.2,
-			innerPadding: 0.1,
+			paddingInner: 0.1,
```
https://d3js.org/d3-scale/band#band_paddingInner

chihsuan committed:
Update docs

chihsuan committed:
Update docs

kangzj commented on projects/js-packages/charts/src/components/bar-list-chart/stories/sample-data.ts:
```diff
@@ -0,0 +1,36 @@
+export const SalesByChannel = [
```
We might want to use lower case to start a variable name except for components.

chihsuan replied:
Ah, right, that helps keep things consistent and clear. I'll rename the variables.

chihsuan replied:
Changed in 7e745d6dd3eb27395337b79da15fa4516f157ed6.

kangzj commented on projects/js-packages/charts/src/components/shared/with-responsive.tsx:
```diff
@@ -17,7 +17,7 @@ type ResponsiveConfig = {
  * @return A functional component that renders the wrapped component with responsive dimensions.
  */
-export function withResponsive< T extends BaseChartProps< unknown > >(
+export function withResponsive< T extends Exclude< BaseChartProps< unknown >, 'options' > >( // 'options' is excluded so that each chart can define its own options type
```
👍

kangzj commented on projects/js-packages/charts/src/components/bar-list-chart/bar-list-chart.tsx:
```diff
@@ -0,0 +1,274 @@
+			valueFormatter: options.valueFormatter ?? ( value => formatNumberCompact( value ) ),
+			labelFormatter: options.labelFormatter ?? ( value => String( value ) ),
+			valuePosition: options.valuePosition ?? width,
+			yOffset: options.yOffset,
```
Guess if we define `defaultYOffset` above this section, we'll be able to use a similar pattern as other values instead of specifying it on L252.

```
chartOptions.yOffset ?? defaultYOffset
```

chihsuan replied:
Good point! I've moved defaultYOffset up above the section like the others in 0aeed47af12eef60c51686cb4a9473ed632308a8.

kangzj reviewed (dismissed):
Great work! It works as described and looks reasonable 👍 

One minor thing I noticed was the texts don't have animations like the bars, which is not a blocker because we might just remove all the animations at the moment. Also it doesn't look too bad.

https://github.com/user-attachments/assets/4f069976-c03c-4fb1-8fa2-3b28e9b8a45f

chihsuan committed:
Refactor sample data exports for BarListChart stories

- Renamed exported constants for sales data from `SalesByChannel` and `SalesByProduct` to `salesByChannel` and `salesByProduct` for consistency with naming conventions.
- Updated references in the BarListChart story files to match the new export names.

chihsuan committed:
Add default y offset calculation for BarListChart

- Introduced a new function `getDefaultYOffset` to calculate the default y offset for the BarListChart based on the data, y scale configuration, height, and whether the chart is multi-series.
- Updated the yOffset property in the chart options to utilize the new function, ensuring proper spacing between bars in multi-series charts.
- Refactored related code for clarity and consistency.

chihsuan committed:
Fix tests
